### PR TITLE
refactor(client): fix types in the nav-links file

### PR DESCRIPTION
**Pull request by Sboonny**
Checklist:

- [x] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [x] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/#/how-to-open-a-pull-request).
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

This is pure refactoring to make working on this file possible, while working I tried to keep Bruce list in mind, I shouldn't have screwed something, but it won't hurt to double check.

Bruce accessibility list.

> Sounds good. I will create a task list with everything that needs to be restored.
> 
> The following are issues that need to be fixed in the main menu/language menu in order to bring it back to proper functionality. You should be able to verify that all of these issues are currently implemented in these menus on production. Unless specified, any current functionality should remain. There may be some things I can't test properly until some of the issues below are implemented, so there is a chance more will be added as things are fixed.
> 
> [ ] When the main navigation menu is expanded and keyboard focus is on an item in main menu, the `Esc` key should close the menu and put keyboard focus on the Menu button.
> 
> [ ] When focus is on the `Change Language` item in the main menu, the up arrow should open the lang menu and place focus on the last item in the lang menu.
> 
> [ ] When focus is on the `Change Language` item in the main menu, the down arrow should open the lang menu and place focus on the first item in the lang menu.
> 
> [ ] When lang menu is opened and keyboard focus is automatically moved into lang menu, the up and down arrow keys should move the keyboard focus up/down one menu item.
> 
> [ ] In the lang menu, when focus is on the last item, arrow down should move focus to first item. Similarly, when focus is on the first item, arrow up should move focus to last item.
> 
> [ ] When focus is on an item in the language menu, `Shift + Tab` should close the lang menu and put focus on the item immediately above the `Change Language` item.
> 
> [ ] When focus is on an item in the language menu, `Tab` should close the lang menu and put focus on item immediately after the `Change Language` item, or if no such item exists then put focus on the next element that can receive focus on the page (which is the Sign In button at this time).
> 
> [ ] When focus is on any item in lang menu, the `Esc` key should close the lang menu and put focus back on the `Change Language` menu item.

### CODE REVIEW

File: client/src/components/Header/components/nav-links.tsx [hunk 1]
```diff
@@ -168,33 +162,38 @@ class NavLinks extends Component<NavLinksProps, NavlinkStates> {
   ): void => {
     const { menuButtonRef, showLanguageMenu, hideMenu } = this.props;
 
-    interface DoKeyPressProp {
-      Escape: () => void;
-      ArrowDown: () => void;
-      ArrowUp: () => void;
-    }
-
     // eslint naming convention should be ignored in key press function, because following the name convention harms accessiblity.
```
Comment (Sboonny) [inline client/src/components/Header/components/nav-links.tsx:165]:
```suggestion
```

**Comment by bbsmooth** [comment]
@Sboonny this looks real good, all the items in my list are working as expected.

Unfortunately, I found two new issues with the menu that are currently present on production.

- When keyboard focus is on "Sign out", `Shift+Tab` takes you to "Night Mode", skipping "Change Language" completely.
- When keyboard focus is on "Sign out", `Esc` moves the focus to "Change Language" instead of closing the menu and moving the focus to the "Menu" button.

Not sure if you want to fix these in this PR or if we should open a separate issue for them?

**Comment by Sboonny** [comment]
I will try 👍. if it's too confusing, I will create a branch from this and move implementing them to second PR

File: client/src/components/Header/components/nav-links.tsx [hunk 2]
```diff
@@ -497,19 +524,17 @@ class NavLinks extends Component<NavLinksProps, NavlinkStates> {
...
-              </button>
-            </li>
-          </Fragment>
+        {currentUserName && (
+          <li className='nav-line' key='sign-out'>
+            <button
+              className='nav-link nav-link-signout'
+              data-value='sign-out-button'
+              onClick={this.handleSignOutClick}
+              onKeyDown={this.handleMenuKeyDown}
```
Comment (Sboonny) [inline client/src/components/Header/components/nav-links.tsx:533]: Luckily, it was simple, it was using the wrong Event.

I had removed the weird Fragments, so I don't really mind adding it, to this PR.

**Review by Nirajn2311 — APPROVED** [review]
LGTM and all the items from @bbsmooth list is implemented

**Review by naomi-lgbt — APPROVED** [review]
Ran through the list, looks like everything is resolved.

**Comment by bbsmooth** [comment]
Wow @Sboonny, you even incorporated the two new bugs I mentioned above!

Unfortunately, a new bug has popped up that is not currently on production, so a small regression.

[ ] If keyboard focus is on the last item in the nav menu, using the tab key should move focus to the next active element on the page (this part works) AND the nav menu should close automatically (this part does now work). This bug is only happening when the user is logged in.

**Comment by Sboonny** [comment]
I have made it that pressing tab hide the menu.

A bit crude because if signout position changed, it won't make sense for it to close the menu, but it's the last element, so I am cheating.

**Review by Nirajn2311 — APPROVED** [review]
Fixes the bug from Bruces latest comment.

**Comment by bbsmooth** [comment]
@Sboonny, it does now close the menu when focus is on the last menu item and you hit the tab key. But it also closes the menu when focus is on the last item and you hit `shift + tab`. :slightly_frowning_face:

**Comment by Sboonny** [comment]
I always like your reviews, it makes me know that I'm misunderstanding concepts or functionality. 👍 I have misunderstood what consideration is, it says it will catch the `shift`, `ctrl` and `window` keypress, and I hoped for more. 🤣 

> The [KeyboardEvent](https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent) interface's key read-only property returns the value of the key pressed by the user, taking into consideration the state of modifier keys such as Shift as well as the keyboard locale and layout.

from https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/key

**Comment by bbsmooth** [comment]
> do you know why are we using role="none" in li elements? 

Because the language menu is implemented as an actual [aria `menu` widget](https://www.w3.org/WAI/ARIA/apg/patterns/menu-button/examples/menu-button-links/#rps2_label) and thus we need to hide the list item's role from screen readers.

**Review by bbsmooth — APPROVED** [review]
Accessibility and functionality looks good. I'm sure you're probably tired of me by now :slightly_smiling_face:

**Comment by Sboonny** [comment]
I didn't lie, when I said I appreciate your reviews 😕

**Pull request merged**